### PR TITLE
[TF FE] Support complex tensors for Equal operation

### PR DESCRIPTION
**Pull request by MonalSD**
### Details:
 - *Extend the loader for the Equal operation on complex type tensors*
 - Addressing #22947

### CODE REVIEW

File: src/frontends/tensorflow_common/src/op/binary_op.cpp [hunk 1]
```diff
@@ -169,6 +169,47 @@ OutputVector translate_addv2_op(const NodeContext& node) {
     return {result};
 }
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/binary_op.cpp:171]: please correct op_table.cpp and make other correction. Check build

File: src/frontends/tensorflow_common/src/op/binary_op.cpp [hunk 2]
```diff
@@ -169,6 +169,47 @@ OutputVector translate_addv2_op(const NodeContext& node) {
...
+    auto y = node.get_input(1);
+
+    auto complex_type_mark_x = as_type_ptr<ComplexTypeMark>(x.get_node_shared_ptr());
+    auto complex_type_mark_y = as_type_ptr<ComplexTypeMark>(y.get_node_shared_ptr());
+
+    // If both inputs are complex tensors
+    if (complex_type_mark_x && complex_type_mark_y) {
+        element::Type complex_part_type_x = complex_type_mark_x->get_complex_part_type();
+        element::Type complex_part_type_y = complex_type_mark_y->get_complex_part_type();
+
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/binary_op.cpp:184]: unused variables

File: src/frontends/tensorflow_common/src/op/binary_op.cpp [hunk 3]
```diff
@@ -169,6 +169,47 @@ OutputVector translate_addv2_op(const NodeContext& node) {
...
+        auto tensor1 = complex_type_mark_x->input_value(0);
+        auto tensor2 = complex_type_mark_y->input_value(0);
+
+        auto equal_op = make_shared<v1::Equal>(tensor1, tensor2);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/binary_op.cpp:188]: check output shape. What shape will be in TF?

**Comment by MonalSD** [comment]
@rkazants addressed the changes, please give it a look.

File: src/frontends/tensorflow_common/src/op/binary_op.cpp (hunk 3 above)
Comment (rkazants) [reply src/frontends/tensorflow_common/src/op/binary_op.cpp:185]: same problem with output shape. What should be done additionally right after equal? Check `Reduce*` ops

**Comment by mlukasze** [comment]
> @rkazants addressed the changes, please give it a look.

not all of them, could you please deliver rest of changes as well?

File: src/frontends/tensorflow_common/src/op/binary_op.cpp (hunk 3 above)
Comment (rkazants) [reply src/frontends/tensorflow_common/src/op/binary_op.cpp:185]: @MonalSD, please think why do you need `ReduceAnd` (check OV opset) operation for this implementation.

Comment (MonalSD) [reply src/frontends/tensorflow_common/src/op/binary_op.cpp:185]: > Reduce
@rkazants I'm not able to understand how reduceAnd operation will be implemented for the equal operation, like are we trying to get the final result as a single element tensor?

Comment (rkazants) [reply src/frontends/tensorflow_common/src/op/binary_op.cpp:185]: Hi @MonalSD, please think why `ReduceAnd` should be used. Equal is element-wise operation and complex numbers are equal when both real and imag parts are equal. So you should reduce boolean tensor by last dimension using `ReduceAnd`.

**Comment by MonalSD** [comment]
@rkazants implemented ReduceLogicalAnd operation. Please give it a look.

File: src/frontends/tensorflow_common/src/op/binary_op.cpp [hunk 4]
```diff
@@ -169,6 +170,37 @@ OutputVector translate_addv2_op(const NodeContext& node) {
...
+    auto complex_type_mark_y = as_type_ptr<ComplexTypeMark>(y.get_node_shared_ptr());
+
+    // If both inputs are complex tensors
+    if (complex_type_mark_x && complex_type_mark_y) {
+        auto tensor1 = complex_type_mark_x->input_value(0);
+        auto tensor2 = complex_type_mark_y->input_value(0);
+
+        auto equal_op = make_shared<v1::Equal>(tensor1, tensor2);
+        // Reduce along the last dimension using ReduceAnd
+        auto reduce_axes = make_shared<ov::Constant>(ov::element::i64, Shape{1}, std::vector<int64_t>{-1});
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/binary_op.cpp:188]: constant header include looks needed

File: src/frontends/tensorflow_common/src/op/binary_op.cpp [hunk 5]
```diff
@@ -169,6 +170,37 @@ OutputVector translate_addv2_op(const NodeContext& node) {
...
+        auto equal_reduced = make_shared<v1::ReduceLogicalAnd>(equal_op, reduce_axes);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/binary_op.cpp:189]: put `keep_dims` flag value explicitly

**Comment by rkazants** [comment]
@MonalSD, also there is build issue.

**Comment by MonalSD** [comment]
@rkazants fixed it please give it a look.

File: src/frontends/tensorflow_common/src/op/binary_op.cpp [hunk 6]
```diff
@@ -169,6 +171,37 @@ OutputVector translate_addv2_op(const NodeContext& node) {
...
 
+OutputVector translate_equal_op(const NodeContext& node) {
+    default_op_checks(node, 2, {"Equal"}, true);
+    auto x = node.get_input(0);
+    auto y = node.get_input(1);
+
+    auto complex_type_mark_x = as_type_ptr<ComplexTypeMark>(x.get_node_shared_ptr());
+    auto complex_type_mark_y = as_type_ptr<ComplexTypeMark>(y.get_node_shared_ptr());
+
+    // If both inputs are complex tensors
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/binary_op.cpp:182]:
```suggestion
    // both inputs are complex tensors
```

File: src/frontends/tensorflow_common/src/op/binary_op.cpp [hunk 7]
```diff
@@ -169,6 +171,37 @@ OutputVector translate_addv2_op(const NodeContext& node) {
...
+    if (complex_type_mark_x && complex_type_mark_y) {
+        auto tensor1 = complex_type_mark_x->input_value(0);
+        auto tensor2 = complex_type_mark_y->input_value(0);
+
+        auto equal_op = make_shared<v1::Equal>(tensor1, tensor2);
+        // Reduce along the last dimension using ReduceAnd
+        auto reduce_axes = make_shared<v0::Constant>(element::i64, Shape{1}, std::vector<int64_t>{-1});
+        auto equal_reduced = make_shared<v1::ReduceLogicalAnd>(equal_op, reduce_axes, true);  // Keepdims set to true
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/binary_op.cpp:190]:
```suggestion
        auto equal_reduced = make_shared<v1::ReduceLogicalAnd>(equal_op, reduce_axes, false);
```

Comment (rkazants) [reply src/frontends/tensorflow_common/src/op/binary_op.cpp:190]: we should not preserve auxiliary dimension.

File: src/frontends/tensorflow_common/src/op/binary_op.cpp [hunk 8]
```diff
@@ -169,6 +171,37 @@ OutputVector translate_addv2_op(const NodeContext& node) {
...
+    auto complex_type_mark_y = as_type_ptr<ComplexTypeMark>(y.get_node_shared_ptr());
+
+    // both inputs are complex tensors
+    if (complex_type_mark_x && complex_type_mark_y) {
+        auto tensor1 = complex_type_mark_x->input_value(0);
+        auto tensor2 = complex_type_mark_y->input_value(0);
+
+        auto equal_op = make_shared<v1::Equal>(tensor1, tensor2);
+        // Reduce along the last dimension using ReduceAnd
+        auto reduce_axes = make_shared<v0::Constant>(element::i64, Shape{1}, std::vector<int64_t>{-1});
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/binary_op.cpp:189]:
```suggestion
        auto reduce_axes = make_shared<v0::Constant>(element::i32, Shape{1}, std::vector<int32_t>{-1});
```

Comment (rkazants) [reply src/frontends/tensorflow_common/src/op/binary_op.cpp:189]: int32 is quite enough for storing -1 value

File: src/frontends/tensorflow_common/src/op/binary_op.cpp [hunk 9]
```diff
@@ -169,6 +171,37 @@ OutputVector translate_addv2_op(const NodeContext& node) {
...
+    auto complex_type_mark_x = as_type_ptr<ComplexTypeMark>(x.get_node_shared_ptr());
+    auto complex_type_mark_y = as_type_ptr<ComplexTypeMark>(y.get_node_shared_ptr());
+
+    // both inputs are complex tensors
+    if (complex_type_mark_x && complex_type_mark_y) {
+        auto tensor1 = complex_type_mark_x->input_value(0);
+        auto tensor2 = complex_type_mark_y->input_value(0);
+
+        auto equal_op = make_shared<v1::Equal>(tensor1, tensor2);
+        // Reduce along the last dimension using ReduceAnd
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/binary_op.cpp:188]:
```suggestion
        // reduce along the last dimension using ReduceAnd
```

File: src/frontends/tensorflow_common/src/op/binary_op.cpp [hunk 10]
```diff
@@ -169,6 +171,37 @@ OutputVector translate_addv2_op(const NodeContext& node) {
...
+
+        auto equal_op = make_shared<v1::Equal>(tensor1, tensor2);
+        // reduce along the last dimension using ReduceAnd
+        auto reduce_axes = make_shared<v0::Constant>(element::i32, Shape{1}, std::vector<int32_t>{-1});
+        auto equal_reduced = make_shared<v1::ReduceLogicalAnd>(equal_op, reduce_axes, false);
+
+        set_node_name(node.get_name(), equal_op);
+
+        auto complex_equal_op =
+            make_shared<ComplexTypeMark>(equal_op, complex_type_mark_x->get_complex_part_type(), element::boolean);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/binary_op.cpp:195]: why is complex mark preserve? TF Equal operation outputs boolean type.

Comment (MonalSD) [reply src/frontends/tensorflow_common/src/op/binary_op.cpp:195]: @rkazants yeah no need to preserve it.

**Comment by MonalSD** [comment]
@rkazants fixed the changes.

File: src/frontends/tensorflow_common/src/op/binary_op.cpp [hunk 11]
```diff
@@ -169,6 +171,36 @@ OutputVector translate_addv2_op(const NodeContext& node) {
...
+        auto tensor2 = complex_type_mark_y->input_value(0);
+
+        auto equal_op = make_shared<v1::Equal>(tensor1, tensor2);
+        // reduce along the last dimension using ReduceAnd
+        auto reduce_axes = make_shared<v0::Constant>(element::i32, Shape{1}, std::vector<int32_t>{-1});
+        auto equal_reduced = make_shared<v1::ReduceLogicalAnd>(equal_op, reduce_axes, false);
+
+        set_node_name(node.get_name(), equal_reduced);
+
+        auto complex_equal_op = make_shared<ComplexTypeMark>(equal_reduced, element::f32, element::boolean);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/binary_op.cpp:194]: why do we need it? Please explain.

Comment (MonalSD) [reply src/frontends/tensorflow_common/src/op/binary_op.cpp]: Apologies for the miss, I have reverted in the next revision.

**Comment by rkazants** [comment]
@MonalSD, please fix build issues. You can find them in CI logs.

File: src/frontends/tensorflow_common/src/op/binary_op.cpp [hunk 12]
```diff
@@ -169,6 +171,34 @@ OutputVector translate_addv2_op(const NodeContext& node) {
...
+
+    // If both inputs are non-complex, perform regular equality comparison
+    auto equal_op = make_shared<v1::Equal>(x, y);
+    set_node_name(node.get_name(), equal_op);
+    return {equal_op};
+}
+
 template OutputVector translate_binary_op<v1::Add>(const NodeContext& node);
 template OutputVector translate_binary_op<v13::BitwiseAnd>(const NodeContext& node);
 template OutputVector translate_binary_op<v13::BitwiseOr>(const NodeContext& node);
```
Comment (rkazants) [inline src/frontends/tensorflow_common/src/op/binary_op.cpp:204]: please remove instantiated template for `Equal` below

File: tests/layer_tests/tensorflow_tests/test_tf_Equal.py [hunk 13]
```diff
@@ -255,3 +255,58 @@ def test_equal_str(self, x_shape, y_shape,
...
+        return tf_net, None
+
+    test_data_complex_equal = [
+        dict(input_shape_1=[2, 6], input_shape_2=[2, 6]),
+        dict(input_shape_1=[2, 4, 5], input_shape_2=[2, 4, 5]),
+        dict(input_shape_1=[1], input_shape_2=[1]),
+    ]
+    
+    @pytest.mark.parametrize("params", test_data_complex_equal)
+    @pytest.mark.precommit_tf_fe
```
Comment (rkazants) [inline tests/layer_tests/tensorflow_tests/test_tf_Equal.py:305]:
```suggestion
    @pytest.mark.precommit
```

**Review by rkazants — COMMENTED** [review]
last comment remained. Other looks good.

**Comment by MonalSD** [comment]
@rkazants please start the build, I have made my changes.

**Comment by mlukasze** [comment]
build_jenkins

**Comment by MonalSD** [comment]
@rkazants please build the run once again...

**Comment by mlukasze** [comment]
build_jenkins

**Comment by mlukasze** [comment]
hey @MonalSD could you poke me on [michal.lukaszewski@intel.com](mailto:michal.lukaszewski@intel.com) or on discord?

**Comment by MonalSD** [comment]
> hey @MonalSD could you poke me on [michal.lukaszewski@intel.com](mailto:michal.lukaszewski@intel.com) or on discord?

Yeah sure...

**Comment by rkazants** [comment]
build_jenkins

**Comment by rkazants** [comment]
@MonalSD, please update duplicating translator in `openvino_tokenizers` https://github.com/openvinotoolkit/openvino_tokenizers/blob/master/src/tensorflow_translators.cpp#L467. Without this update, the test will fail

**Comment by MonalSD** [comment]
> @MonalSD, please update duplicating translator in `openvino_tokenizers` https://github.com/openvinotoolkit/openvino_tokenizers/blob/master/src/tensorflow_translators.cpp#L467 by creating separate PR to that repository. Without this update, the test will fail
> 
> Also, fix build issues reported in CI

hey, should I add the support for complex type there?

**Comment by rkazants** [comment]
> > @MonalSD, please update duplicating translator in `openvino_tokenizers` https://github.com/openvinotoolkit/openvino_tokenizers/blob/master/src/tensorflow_translators.cpp#L467 by creating separate PR to that repository. Without this update, the test will fail
> > Also, fix build issues reported in CI
> 
> hey, should I add the support for complex type there?

yes, please create separate PR to https://github.com/openvinotoolkit/openvino_tokenizers. And update https://github.com/openvinotoolkit/openvino_tokenizers/blob/master/src/tensorflow_translators.cpp#L475.

We should merge PR at that repository first and only this one after.

**Comment by MonalSD** [comment]
hey @rkazants raised the PR as you mentioned https://github.com/openvinotoolkit/openvino_tokenizers/pull/122

**Pull request closed without merging**